src/py21cmmc/emulator.py

```diff
@@ -0,0 +1,196 @@
+        self.Ts_err = all_emulator_numbers['Ts_err']
+        self.xHI_err = all_emulator_numbers['xHI_err']
+        self.tau_err = all_emulator_numbers['tau_err']
+        log.info('Finished emulator init sucessfully.')
+
+    def predict(self, params):
+        r"""
+        Call the emulator, evaluate it at the given parameters, restore dimensions.
+    
+        Parameters
+        ----------
+        theta : p21.AstroParams or np.ndarray
+        An array with the nine astro_params input all $\in [0,1]$.
+        """
+        log.info('Beginning prediction...')
+        if isinstance(params, p21.AstroParams) or type(params) == dict:
+            log.info('Must convert input params from AstroParams or dict into array.')
+            param_keys = ['F_STAR10','ALPHA_STAR','F_ESC10','ALPHA_ESC','M_TURN', 
+               't_STAR','L_X','NU_X_THRESH','X_RAY_SPEC_INDEX']
+            theta = np.zeros(len(param_keys))
+            for i, key in enumerate(param_keys):
+                if isinstance(params, p21.AstroParams):
+                    theta[i] = params.defining_dict[key]
+                else:
+                    theta[i] = params[key]
+        elif type(params) == np.ndarray:
+            theta = params.copy()
+        else:
+            raise TypeError('theta is in the wrong format. Should be AstroParams object or nine astrophysical parameters in same order as param_keys.')
+        if len(theta.shape) == 1:
+            theta = theta.reshape([1,-1])
+        normed = True
+        # Check that theta is normalized, if not, normalise it.
+        if isinstance(params, p21.AstroParams) or max(theta.ravel()) > 1 or min(theta.ravel()) < 0:
+            log.info('Theta is not normalized.')
+            normed = False # to indicate that input params was not normalised
+            for i in range(theta.shape[-1]):
+                if i == 0 or i == 2 or i == 4 or i == 6:
+                    theta[:,i] = np.log10(theta[:,i])
+                if i == 7:
+                    theta[:,i] /= 1000. #E0 in keV
+                theta[:,i] -= self.limits[i,0]
+                theta[:,i] /= (self.limits[i,1] - self.limits[i,0])
+        log.info('Normed params: %s' % theta)
+        log.info('Start emulator prediction...')
+        emu_pred = self.model.predict(theta, verbose = False)
+        log.info('End emulator prediction.')
+
+        Tb_pred_normed = emu_pred[:,:84]
+        xHI_pred = emu_pred[:,84:84*2]
+        Ts_pred_normed = emu_pred[:,2*84:84*3]
+        Ts_undefined_pred = emu_pred[:,84*3]
+        PS_pred_normed = emu_pred[:,84*3+1:].reshape((theta.shape[0], 60,12))
+
+        # Set the xHI < z(Ts undefined) to 0
+        xHI_pred_fix = np.zeros(xHI_pred.shape)
+
+        tau = np.zeros(theta.shape[0])
+        uvlfs = np.zeros((theta.shape[0], 3, len(self.uv_lf_zs),100))
+        for i in range(theta.shape[0]):
+            zbin = np.argmin(abs(self.zs - Ts_undefined_pred[i]))
+            if xHI_pred[i,zbin] < 1e-1:
```

Comment:
I dont understand what this part is about...

Reply:
The emulator predicts the redshift at which Ts becomes undefined. This is also approx the same redshift at which xHI becomes 0, so this part says if the xHI value at this bin where Ts is not defined anymore is less than 0.1, then set every xHI at lower z to 0.